Add pyeCMD, the Pythonic eCMD wrapper

fenkes-ibm opened the pull request:
Add an eCMD wrapper module that is mostly generated automatically from Doxygen and will apply a set of Pythonic transformations to the more C-like pyapi -- exceptions, output parameters as function return values, functions that apply to targets are now Target methods, and several more. Please see the initial comment of __init__.py for details of the transformations.
Incidentally update pyapi to map all string and enum output parameters into return values, hopefully making all eCMD functions available via Python now.

mklight commented on config.py:
```diff
@@ -157,6 +157,10 @@
 optgroup.add_argument("--without-python3", action='store_true', help="Disable python3 module build\n"
                                                                      "CREATE_PY3API from the environment")
 
+# --without-python3
```
update comment to --without-pyecmd

mklight commented on ecmd-core/pyapi/ecmdClientPyapi.i:
```diff
@@ -76,6 +75,7 @@
 /*********** Start Files to swigify ***********/
 %include "ecmdDefines.H"
 %include "ecmdStructs.H"
+%apply std::string &OUTPUT { std::string & };    // Apply after ecmdStructs.H because this apply is too broad for ecmdStructs and will transform input parameters
```
I see this modifying input and input/output string reference variables in functions like the following are being interpreted by this:

```
ecmdQueryScomGroup(..., std::string & io_scomGroupFileVersion)
getScomGroup(..., std::string & io_scomGroupFileVersion)
ecmdQueryFileLocation(..., std::string & io_version)
ecmdQueryBlock(..., const std::string & i_metadataHWPName = "all")
ecmdGetBlock(..., const std::string & i_metadataBlockEntry, ...)
makeSPSystemCall(..., const std::string & i_command, ...)
...
```

mklight commented on ecmd-core/pyapi/ecmdClientPyapi.i:
```diff
@@ -12,13 +12,12 @@
 /*********** Start Applies ***********/
 // These are used to map C types that swig doesn't understand to types swig does understand
+%apply uint32_t &OUTPUT { uint32_t & };
+%apply int &OUTPUT { enum SWIGTYPE & };
```
This is affecting this poorly defined functions like
ecmdGetClockSpeedSensor(..., ecmdClockSpeedType_t & i_clockspeedUnit, ...)

fenkes-ibm commented:
Thanks for the review! I changed the comment and updated the pyapi commit to filter for o_* arguments only. It became a bit more complex but looks like it's producing the correct output now.
There are a few cases left where io_* argument should really be both input parameters _and_ a return value due to being immutable, but fixing those would go beyond the scope of this pull request (and they didn't work before either). I'd rather tackle them in a separate go.

mklight commented on ecmd-core/pyapi/makefile:
```diff
@@ -116,6 +116,15 @@ ${SOURCE_OBJS} ${SWIG_OBJS}: ${OBJPATH}%.o : %.C ${INCLUDES}
 	@echo Swigging $@
 	${VERBOSE}${SWIG} ${SWIGFLAGS} ${DEFINES} ${PYDEFINES} -o ${SRCPATH}/${SWIG_SOURCE} -outdir ${OUTPY} ../pyapi/ecmdClientPyapi.i
+	@echo Generating mapping of output parameters
+	${VERBOSE}{ OUTARGS=$$(grep -Eo '\Wo_\w+' ${OUTPY}/${PY_TARGET} | cut -c2- | sort | uniq); \
```
I was going through the generated output again and seeing some weirdness with APIs like ecmdGpioReadPin().
I added a line to this makefile to output the results of the grep and it is not finding any matches from this search

```
    grep -Eo '\Wo_\w+' ${OUTPY}/${PY_TARGET}
```

fenkes-ibm commented:
Huh, looks like not all greps are created equal. I re-implemented the %apply generation in Python and fixed a build issue too (where the build failed if make doxygen had never been run).

fenkes-ibm commented:
I understand now. Your build system is using SWIG 3.0.2 which does not emit parameter names into the generated Python module but uses a simple "*args" catch-all parameter, so my tooling cannot extract parameter names from there.
I changed the makefile to extract parameter names from the header files instead, and use SWIG's dependency extraction to find all those header files. Tested with SWIG 3.0.2 and it nets me the expected C code for ecmdGpioReadPin().
This will be more stable too as it doesn't depend on the SWIG generated code.

mklight commented:
@fenkes-ibm I am still reviewing build results. It looks to fix the issue I saw.

mklight commented on ecmd-core/pyecmd/__init__.py:
```diff
@@ -0,0 +1,236 @@
+# IBM Confidential
```
@fenkes-ibm could you update these prologs to be similar to what is in ecmdClientCapi.H instead of the IBM Confidential text.

mklight commented:
@fenkes-ibm Ready to merge with prolog updates. Tests I ran looked good.

mklight commented:
@fenkes-ibm Thanks for submitting this change.
